src/libgrate/matrix.c

```diff
@@ -143,7 +143,7 @@ void mat4_perspective(struct mat4 *m, float fov, float aspect,
 	m->xx = cotangent / aspect;
 	m->yy = cotangent;
 	m->zz = -(far + near) / depth;
-	m->zw = -1.0f;
-	m->wz = -2.0f * near * far / depth;
+	m->zw = -2.0f * near * far / depth;
+	m->wz = -1.0f;
```

Comment:
Yeah, I think we've just doing the wrong memory-layout for our matrices.

https://github.com/digetx/grate/blob/7ff314d7b7aa37a69a425232f1089a87463f218a/src/libgrate/matrix.c#L72

This corresponds to the fourth float:

https://github.com/digetx/grate/blob/7ff314d7b7aa37a69a425232f1089a87463f218a/src/libgrate/matrix.h#L29

However, it should be the twelfth float:

https://github.com/mesa3d/mesa/blob/master/src/mesa/math/m_matrix.c#L1098